## 1. Prepare reference database

The next step is to obtain a reference database of 18S sequences for phytoplankton. This can be done by searching public databases such as NCBI GenBank, PR2 or SILVA, or by creating a custom database using reference sequences obtained from previous studies. In this case, we use PR2 directly.

Download the PR2 reference database in fasta format from the PR2 website and save it as "pr2.fasta" in the current directory.

```
cd ~/hdd/MetaBarcoding2023/reference_database
wget https://github.com/pr2database/pr2database/releases/download/v4.13.0/pr2_version_4.13.0_18S_mothur.fasta.gz
gunzip pr2_version_4.13.0_18S_mothur.fasta.gz -c > pr2.fasta

wget https://github.com/pr2database/pr2database/releases/download/v4.13.0/pr2_version_4.13.0_18S_mothur.tax.gz
gunzip pr2_version_4.13.0_18S_mothur.tax.gz -c > pr2_tax.txt
```

The filtered reads can be aligned to the reference database using QIIME2, but before that we need to create the reference index from the PR2 reference database.
Because the QIIME2 is very picky on the dependencies, we create a dedicated conda enironment for QIIME2:

```
wget https://data.qiime2.org/distro/core/qiime2-2022.11-py38-linux-conda.yml
conda env create -n qiime2-2022.11 --file qiime2-2022.11-py38-linux-conda.yml
rm qiime2-2022.11-py38-linux-conda.yml
conda activate qiime2-2022.11
```

Now import the PR2 reference database into QIIME2

```
conda activate qiime2-2022.11

qiime tools import \
  --type 'FeatureData[Sequence]' \
  --input-path $PWD/pr2.fasta \
  --output-path $PWD/pr2.qza

qiime tools import \
  --type 'FeatureData[Taxonomy]' \
  --input-format HeaderlessTSVTaxonomyFormat \
  --input-path $PWD/pr2_tax.txt \
  --output-path $PWD/pr2_tax.qza
```

## 2.Alignment and taxonomic assignment

Before the alignment, we need to import all the reads files with the same barcode into one QIIME2 recongnized file. This will also cover the merge processes we usually would do.

```
for dir in ~/hdd/MetaBarcoding2023/basecalled_reads/pass/barcode*/
do
  barcode=$(basename "$dir")
  echo -e "sample-id\tabsolute-filepath\tdirection" > "$dir/manifest"
  
  for file in "$dir"*.fastq.gz
  do
    filename=$(basename "$file")
    echo -e "${filename%.fastq.gz}\t$file\tforward" >> "$dir/manifest"
  done
  
  qiime tools import \
    --type 'SampleData[SequencesWithQuality]' \
    --input-path "$dir/manifest" \
    --output-path "$dir../../../merged_qza/${barcode}.qza" \
    --input-format SingleEndFastqManifestPhred33V2
done
```
Now finally we run this QIIME2 pipeline to do the alignment and taxonomic assignment.

```
# Set input and output directories and reference database directory
input_dir=~/hdd/MetaBarcoding2023/merged_qza
output_dir=~/hdd/MetaBarcoding2023/asv
ref_dir=~/hdd/MetaBarcoding2023/reference_database

# Loop through all files in the input directory that start with "barcode"
for file in $input_dir/barcode*.qza
do
  filename=$(basename "$file")

  # Run the dada2 denoise-single command on the current file,it's used for quality filtering, denoising, and chimera removal of single-end sequence data.
  qiime dada2 denoise-single \
    --i-demultiplexed-seqs $file \
    --p-n-threads 20 \
    --p-trunc-len 0 \
    --o-table $output_dir/${filename%.qza}_asv_table.qza \
    --o-representative-sequences $output_dir/${filename%.qza}_rep_seqs.qza \
    --o-denoising-stats $output_dir/${filename%.qza}_stats_dada2.qza
    
  # Run the feature-classifier classify-consensus-vsearch command on the rep_seqs output from the previous command
  qiime feature-classifier classify-consensus-vsearch \
    --i-query $output_dir/${filename%.qza}_rep_seqs.qza \
    --i-reference-reads $ref_dir/pr2.qza \
    --i-reference-taxonomy $ref_dir/pr2_tax.qza \
    --p-perc-identity 0.95 \
    --p-threads 20 \
    --o-classification $output_dir/${filename%.qza}_vsearch_classification.qza \
    --o-search-results $output_dir/${filename%.qza}_vsearch_results.qza
    
  # Export the ASV table from the previous command to a TSV file
  qiime tools export \
    --input-path $output_dir/${filename%.qza}_asv_table.qza \
    --output-path $output_dir/${filename%.qza}_asv_table
  biom convert -i $output_dir/${filename%.qza}_asv_table/feature-table.biom -o $output_dir/${filename%.qza}_asv_table.tsv --to-tsv

  # Export the vsearch classification from the previous command to a TSV file
  qiime tools export --input-path $output_dir/${filename%.qza}_vsearch_classification.qza --output-path $output_dir/${filename%.qza}_asv_table
done
```
Here are the explainations for each commandï¼š

1. The `dada2 denoise-single` command requires the input data to be in the demultiplexed sequence format, which contains reads that have already been assigned to their respective samples based on the barcode. It performs quality filtering by removing reads with low-quality bases and trimming the reads to a fixed length specified by the user. The output of the `dada2 denoise-single` command is a feature table, which is a matrix of the frequency of each ASV in each sample, and a representative sequence file, which contains one representative sequence for each ASV. These outputs can be used for downstream analysis, such as alpha and beta diversity analysis, taxonomic classification, and functional inference.
2. `qiime feature-classifier classify-consensus-vsearch` is a command used to classify the taxonomic identity of each ASV in a feature table by comparing its sequence to a reference database. The command requires three input files: the representative sequences of the ASVs, the reference sequences, and the reference taxonomy file. It compares each query sequence to the reference sequences using a pairwise alignment, and calculates a consensus taxonomy for each ASV based on the taxonomy of the reference sequences that match the query sequence with a specified percent identity threshold (here we use 0.97). The consensus taxonomy is determined based on the most common taxonomic assignment of the matching reference sequences. The output are a classification file, which contains the taxonomic assignment of each ASV in the feature table, as well as a search results file, which contains detailed information about the search results for each query sequence.
3. `qiime tools export` and `biom convert` commands will export the .qza files into a readable table which will be used in the downstream analysis.
