## 1. Prepare reference database

The next step is to obtain a reference database of 18S sequences for phytoplankton. This can be done by searching public databases such as NCBI GenBank, PR2 or SILVA, or by creating a custom database using reference sequences obtained from previous studies. In this case, we use SILVA directly.

Download the latest reference database from the SILVA website ( https://www.arb-silva.de/download/ ) in FASTA format NOT arb format!

```
mkdir ~/hdd/MetaBarcoding2023/reference_database
cd ~/hdd/MetaBarcoding2023/reference_database
wget https://www.arb-silva.de/fileadmin/silva_databases/current/Exports/SILVA_138.1_SSURef_NR99_tax_silva.fasta.gz
```

The alignment will be done by VSEARCH and the taxonomic assignment will be done by QIIME2, but before that we need to create the reference index from the reference database. Because the QIIME2 is very picky on the dependencies, we create a dedicated conda enironment for QIIME2:

```
wget https://data.qiime2.org/distro/core/qiime2-2022.11-py38-linux-conda.yml
conda env create -n qiime2-2022.11 --file qiime2-2022.11-py38-linux-conda.yml
rm qiime2-2022.11-py38-linux-conda.yml
conda activate qiime2-2022.11
```

Index the reference database for fast search:

```
vsearch --makeudb_usearch SILVA_138.1_SSURef_NR99_tax_silva.fasta.gz \
--output SILVA_138.1_SSURef_NR99_tax_silva.udb
```
## 2.Alignment

## 2.1.Clustering before alignment (optional)

Due to the poor quality of the Nanopore reads, we decided to perform a clustering before the alignment, so the alignment rate can be improved.

```
# Set the input and output directories
input_dir="/MetaBarcoding2023/merged/"
output_dir="/MetaBarcoding2023/clustered/"

# Loop through each file in the input directory
for file in "$input_dir"*.fastq.gz
do
    # Get the file name without the path and extension
    file_name=$(basename "$file" .fastq.gz)
    
    # Convert fastq.gz to fasta format
    gunzip -c "$file" | sed -n '1~4s/^@/>/p;2~4p' > "$output_dir""$file_name".fasta
    
    # Cluster the reads using VSEARCH
    vsearch --cluster_fast "$output_dir""$file_name".fasta --id 0.9 --msaout --uc "$output_dir""$file_name"_cluster.uc --centroids "$output_dir""$file_name"_centroid.fasta --relabel "OTU_" --threads 16
done
```

The `vsearch --usearch_global` function only accept fasta file, so you need to convert fastq.gz into fasta files before input. 

```
# Convert FASTQ files to FASTA format
for f in ~/hdd/MetaBarcoding2023/merged/*.fastq.gz
do
  base=$(basename $f .fastq.gz)
  gunzip -c $f | sed -n '1~4s/^@/>/p;2~4p' > ~/hdd/MetaBarcoding2023/merged/$base.fasta
done
```

Perform sequence alignment of the reads to the reference database in batch

```
for file in ~/hdd/MetaBarcoding2023/merged/*.fasta; do
    base=$(basename $file .fasta)
    vsearch --usearch_global $file \
    --db ~/hdd/MetaBarcoding2023/reference_database/SILVA_138.1_SSURef_NR99_tax_silva.udb \
    --id 0.97 \
    --strand plus \
    --matched ~/hdd/MetaBarcoding2023/aligned/$base.aligned.fasta \
    --notmatched /MetaBarcoding2023/aligned/$base.unaligned.fasta
done
```

## 3.Taxonomic assignment
