## 1. Prepare reference database

The next step is to obtain a reference database of 18S sequences for phytoplankton. This can be done by searching public databases such as NCBI GenBank, PR2 or SILVA, or by creating a custom database using reference sequences obtained from previous studies. In this case, we use PR2 directly.

Download the PR2 reference database in fasta format from the PR2 website and save it as "pr2.fasta" in the current directory.

```
cd ~/hdd/MetaBarcoding2023/reference_database
wget https://github.com/pr2database/pr2database/releases/download/v4.14.0/pr2_version_4.14.0_SSU_taxo_long.fasta.gz
gunzip pr2_version_4.14.0_SSU_taxo_long.fasta.gz -c > pr2.fasta
```

The filtered reads can be aligned to the reference database using Minimap2, but before that we need to create the minimap2 index from the PR2 reference database, install minimap2 if not already installed.

```
minimap2 -d pr2.mmi pr2.fasta
```

You will have a file `pr2.mmi` in current directory and see something like this in the console if everything goes well:

```
zixia@Rstudio:~/hdd/MetaBarcoding2023/reference_database$ minimap2 -d pr2.mmi pr2.fasta
[M::mm_idx_gen::6.950*1.77] collected minimizers
[M::mm_idx_gen::8.780*2.02] sorted minimizers
[M::main::10.281*1.87] loaded/built the index for 197602 target sequence(s)
[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 197602
[M::mm_idx_stat::10.404*1.86] distinct minimizers: 2686938 (54.99% are singletons); average occurrences: 19.292; average spacing: 5.431; total length: 281544243
[M::main] Version: 2.24-r1122
[M::main] CMD: minimap2 -d pr2.mmi pr2.fasta
[M::main] Real time: 10.553 sec; CPU: 19.548 sec; Peak RSS: 1.553 GB
```

## 2.Alignment

Because the QIIME2 is very picky on the dependencies, we create a dedicated conda enironment for QIIME2:

```
wget https://data.qiime2.org/distro/core/qiime2-2022.11-py38-linux-conda.yml
conda env create -n qiime2-2022.11 --file qiime2-2022.11-py38-linux-conda.yml
rm qiime2-2022.11-py38-linux-conda.yml
conda activate qiime2-2022.11
```

Import the reads files into QIIME2:

```
conda activate qiime2-2022.11

for dir in ~/hdd/MetaBarcoding2023/basecalled_reads/pass/barcode*/
do
  barcode=$(basename "$dir")
  echo -e "sample-id\tabsolute-filepath\tdirection" > "$dir/manifest"
  
  for file in "$dir"*.fastq.gz
  do
    filename=$(basename "$file")
    echo -e "${filename%.fastq.gz}\t$file\tforward" >> "$dir/manifest"
  done
  
  qiime tools import \
    --type 'SampleData[SequencesWithQuality]' \
    --input-path "$dir/manifest" \
    --output-path "$dir../../../merged_qza/${barcode}.qza" \
    --input-format SingleEndFastqManifestPhred33V2
done
```

<details>
  <summary>Click to expand</summary>
  
  
</details>

## Taxonomic classification

The aligned reads can be assigned taxonomic classifications based on the reference database using R packages `DECIPHER` and `phyloseq`. The following analysis ara carried out in R console or Rstudio.

```

```


