## 1. Prepare reference database

The next step is to obtain a reference database of 18S sequences for phytoplankton. This can be done by searching public databases such as NCBI GenBank, PR2 or SILVA, or by creating a custom database using reference sequences obtained from previous studies. In this case, we use PR2 directly.

Download the PR2 reference database in fasta format from the PR2 website and save it as "pr2.fasta" in the current directory.

```
cd ~/hdd/MetaBarcoding2023/reference_database
wget https://github.com/pr2database/pr2database/releases/download/v4.13.0/pr2_version_4.13.0_18S_mothur.fasta.gz
gunzip pr2_version_4.13.0_18S_mothur.fasta.gz -c > pr2.fasta

wget https://github.com/pr2database/pr2database/releases/download/v4.13.0/pr2_version_4.13.0_18S_mothur.tax.gz
gunzip pr2_version_4.13.0_18S_mothur.tax.gz -c > pr2_tax.txt
```

The filtered reads can be aligned to the reference database using QIIME2, but before that we need to create the reference index from the PR2 reference database.
Because the QIIME2 is very picky on the dependencies, we create a dedicated conda enironment for QIIME2:

```
wget https://data.qiime2.org/distro/core/qiime2-2022.11-py38-linux-conda.yml
conda env create -n qiime2-2022.11 --file qiime2-2022.11-py38-linux-conda.yml
rm qiime2-2022.11-py38-linux-conda.yml
conda activate qiime2-2022.11
```

Now import the PR2 reference database into QIIME2

```
conda activate qiime2-2022.11

qiime tools import \
  --type 'FeatureData[Sequence]' \
  --input-path $PWD/pr2.fasta \
  --output-path $PWD/pr2.qza

qiime tools import \
  --type 'FeatureData[Taxonomy]' \
  --input-format HeaderlessTSVTaxonomyFormat \
  --input-path $PWD/pr2_tax.txt \
  --output-path $PWD/pr2_tax.qza
```

## 2.Alignment

Import the reads files into QIIME2:

```
for dir in ~/hdd/MetaBarcoding2023/basecalled_reads/pass/barcode*/
do
  barcode=$(basename "$dir")
  echo -e "sample-id\tabsolute-filepath\tdirection" > "$dir/manifest"
  
  for file in "$dir"*.fastq.gz
  do
    filename=$(basename "$file")
    echo -e "${filename%.fastq.gz}\t$file\tforward" >> "$dir/manifest"
  done
  
  qiime tools import \
    --type 'SampleData[SequencesWithQuality]' \
    --input-path "$dir/manifest" \
    --output-path "$dir../../../merged_qza/${barcode}.qza" \
    --input-format SingleEndFastqManifestPhred33V2
done
```

```
input_dir=~/hdd/MetaBarcoding2023/merged_qza
output_dir=~/hdd/MetaBarcoding2023/asv
ref_dir=~/hdd/MetaBarcoding2023/reference_database

for file in $input_dir/barcode*.qza
do
  filename=$(basename "$file")

  # Run the dada2 denoise-single command
  qiime dada2 denoise-single \
    --i-demultiplexed-seqs $file \
    --p-n-threads 20 \
    --p-trunc-len 0 \
    --o-table $output_dir/${filename%.qza}_asv_table.qza \
    --o-representative-sequences $output_dir/${filename%.qza}_rep_seqs.qza \
    --o-denoising-stats $output_dir/${filename%.qza}_stats_dada2.qza
    
  qiime feature-classifier classify-consensus-vsearch \
    --i-query $output_dir/${filename%.qza}_rep_seqs.qza \
    --i-reference-reads $ref_dir/pr2.qza \
    --i-reference-taxonomy $ref_dir/pr2_tax.qza \
    --p-perc-identity 0.95 \
    --p-threads 20 \
    --o-classification $output_dir/${filename%.qza}_vsearch_classification.qza \
    --o-search-results $output_dir/${filename%.qza}_vsearch_results.qza
	
	
  qiime tools export \
    --input-path $output_dir/${filename%.qza}_asv_table.qza \
    --output-path $output_dir/${filename%.qza}_asv_table

  biom convert -i $output_dir/${filename%.qza}_asv_table/feature-table.biom -o $output_dir/${filename%.qza}_asv_table.tsv --to-tsv

  qiime tools export --input-path $output_dir/${filename%.qza}_vsearch_classification.qza --output-path $output_dir/${filename%.qza}_asv_table
done
```

```


```

<details>
  <summary>Click to expand</summary>
  
  
</details>

## Taxonomic classification

The aligned reads can be assigned taxonomic classifications based on the reference database using R packages `DECIPHER` and `phyloseq`. The following analysis ara carried out in R console or Rstudio.

```

```


